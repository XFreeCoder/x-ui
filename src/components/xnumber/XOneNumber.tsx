import React, { FC, useMemo } from 'react'
import { animated, useSpring } from 'react-spring'
import { usePreviousDistinct } from 'react-use'
import { isEqual } from 'lodash'
import { OneNumType, DirectionType, IXOneNumberProps } from './interface'
import XOneChar from './XOneChar'

const AnimatedXOneChar = animated(XOneChar)

const getNext = (from: OneNumType, direction: DirectionType): OneNumType => {
  if (direction === 'up') {
    return ((from + 1) % 10) as OneNumType
  }
  if (from === 0) {
    return 9
  }
  return (from - 1) as OneNumType
}

const getRange = (
  from: OneNumType,
  to: OneNumType,
  direction: DirectionType
) => {
  if (from === to) {
    return [to]
  }
  const range: OneNumType[] = []
  let cur = from
  do {
    cur = getNext(cur, direction)
    range.push(cur)
  } while (cur !== to)
  return range
}

const XOneNumber: FC<IXOneNumberProps> = (props) => {
  const { value = 0, direction = 'up', ...restProps } = props
  const preValue = usePreviousDistinct(value) || 0
  const values = useMemo(() => getRange(preValue, value, direction), [
    value,
    preValue,
    direction
  ])
  const { index } = useSpring({
    index: values.length - 1,
    from: { index: 0 },
    reset: true
  })
  return (
    <AnimatedXOneChar
      value={index.interpolate((i) => values[Math.floor(i)])}
      direction={direction}
      {...restProps}
    />
  )
}

export default React.memo(XOneNumber, (preProps, nextProps) => {
  const { direction: preDirection, ...restPreProps } = preProps
  const { direction: nextDirection, ...restNextProps } = nextProps
  // only direction change, dot re-render
  return isEqual(restPreProps, restNextProps)
})
