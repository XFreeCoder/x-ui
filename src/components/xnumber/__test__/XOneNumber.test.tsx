import React from 'react'
import { render } from '@testing-library/react'
import XOneNumber from '../XOneNumber'
import { animationConfig } from '../../../constants'

test('render correctly', async () => {
  const container = render(
    <XOneNumber value={5} animationConfig={animationConfig.default} />
  )
  expect(container).toMatchSnapshot()
})
