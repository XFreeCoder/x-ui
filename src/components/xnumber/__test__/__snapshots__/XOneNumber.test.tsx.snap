// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="sc-bdnylx dQYGTz"
        style="height: 0px; width: 0px;"
      >
        <span
          style="visibility: hidden;"
        >
          5
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 1; transform: translateY(0px);"
        >
          0
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          1
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          2
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          3
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          4
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          5
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          6
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          7
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          8
        </span>
        <span
          class="sc-gtssRu huSdcX"
          style="opacity: 0; transform: translateY(0px);"
        >
          9
        </span>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="sc-bdnylx dQYGTz"
      style="height: 0px; width: 0px;"
    >
      <span
        style="visibility: hidden;"
      >
        5
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 1; transform: translateY(0px);"
      >
        0
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        1
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        2
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        3
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        4
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        5
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        6
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        7
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        8
      </span>
      <span
        class="sc-gtssRu huSdcX"
        style="opacity: 0; transform: translateY(0px);"
      >
        9
      </span>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
